server/src/main/java/org/elasticsearch/index/engine/TranslogOperationAsserter.java

```diff
@@ -0,0 +1,127 @@
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ *
+ * A utility class to assert that translog operations with the same sequence number
+ * in the same generation are either identical or equivalent when synthetic sources are used.
+ */
+public abstract class TranslogOperationAsserter {
+    public static final TranslogOperationAsserter DEFAULT = new TranslogOperationAsserter() {
+    };
+
+    private TranslogOperationAsserter() {
+
+    }
+
+    public static TranslogOperationAsserter withEngineConfig(EngineConfig engineConfig) {
+        return new TranslogOperationAsserter() {
+            @Override
+            boolean sameIndexOperation(Translog.Index o1, Translog.Index o2) throws IOException {
+                if (super.sameIndexOperation(o1, o2)) {
+                    return true;
+                }
+                if (engineConfig.getIndexSettings().isRecoverySourceSyntheticEnabled()) {
+                    return super.sameIndexOperation(synthesizeSource(engineConfig, o1), o2)
+                        || super.sameIndexOperation(o1, synthesizeSource(engineConfig, o2));
+                }
+                return false;
+            }
+        };
+    }
+
+    static Translog.Index synthesizeSource(EngineConfig engineConfig, Translog.Index op) throws IOException {
+        final ShardId shardId = engineConfig.getShardId();
+        MappingLookup mappingLookup = engineConfig.getMapperService().mappingLookup();
+        DocumentParser documentParser = engineConfig.getMapperService().documentParser();
+        try (var reader = new TranslogDirectoryReader(shardId, op, mappingLookup, documentParser, engineConfig, () -> {})) {
+            final Engine.Searcher searcher = new Engine.Searcher(
+                "assert_translog",
+                reader,
+                engineConfig.getSimilarity(),
+                engineConfig.getQueryCache(),
+                engineConfig.getQueryCachingPolicy(),
+                () -> {}
+            );
+            try (
+                LuceneSyntheticSourceChangesSnapshot snapshot = new LuceneSyntheticSourceChangesSnapshot(
+                    mappingLookup,
+                    searcher,
+                    LuceneSyntheticSourceChangesSnapshot.DEFAULT_BATCH_SIZE,
+                    Integer.MAX_VALUE,
+                    op.seqNo(),
+                    op.seqNo(),
+                    true,
+                    false,
+                    engineConfig.getIndexSettings().getIndexVersionCreated()
+                )
+            ) {
+                final Translog.Operation normalized = snapshot.next();
+                assert normalized != null : "expected one operation; got zero";
+                return (Translog.Index) normalized;
+            }
+        }
+    }
+
+    boolean sameIndexOperation(Translog.Index o1, Translog.Index o2) throws IOException {
+        // TODO: We haven't had timestamp for Index operations in Lucene yet, we need to loosen this check without timestamp.
```

Comment:
This todo confuses me a little. I notice the auto-gen-timestamp is not included, I assume because we do not hold it in lucene, which is fair. But then what is the todo about?

Reply:
I copied this TODO. I have removed it.